release-21.1: catalog: add telemetry for descriptor validation errors

postamar opened the pull request:
Backport 1/1 commits from #62546.

/cc @cockroachdb/release

---

This commit adds `sql.schema.validation_errors.*` telemetry keys to
descriptor validation errors.

Fixes #61786.

Release note: None

Commit:
catalog: add telemetry for descriptor validation errors

This commit adds `sql.schema.validation_errors.*` telemetry keys to
descriptor validation errors.

Fixes #61786.

Release note: None

Commit:
catalog: fix descriptor validation early exit

This patch fixes a bug in doctor and crdb_internal.invalid_objects
preventing multiple validation errors originating from different
validation levels from being properly reported.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/62882)
<!-- Reviewable:end -->

postamar commented:
This backport includes the commit in https://github.com/cockroachdb/cockroach/pull/62878 which fixes a regression introduced by 3316a6f.

Commit:
sql: improve telemetry test cases for descriptor corruption

This patch improves the deliberate corruption of the descriptor state in
the `error` suite of the telemetry logic tests.

Release note: None

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 12 of 12 files at r1, 2 of 2 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/62882#-:-MXbxnb8D-pc4UhRVb5E:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @postamar)



<!-- Sent from Reviewable.io -->
